src/main/java/com/ss/mqtt/broker/service/Subscriber.java

```diff
@@ -0,0 +1,35 @@
+package com.ss.mqtt.broker.service;
+
+import com.ss.mqtt.broker.model.QoS;
+import com.ss.mqtt.broker.model.SubscribeRetainHandling;
+import com.ss.mqtt.broker.network.MqttClient;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+import java.util.Objects;
+
+@Getter
+@Builder
+@RequiredArgsConstructor
+public class Subscriber {
+
+    private final MqttClient mqttClient;
+    private final QoS qos;
+    private final SubscribeRetainHandling retainHandling;
+    private final boolean noLocal;
+    private final boolean retainAsPublished;
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Subscriber that = (Subscriber) o;
+        return Objects.equals(mqttClient, that.mqttClient);
```

Comment:
the same client can be indifferent subscribers with different settings, you need to use all fields to compare it

Reply:
http://docs.oasis-open.org/mqtt/mqtt/v5.0/cs02/mqtt-v5.0-cs02.html#_Toc514345442

> If a Server receives a SUBSCRIBE packet containing a Topic Filter that is identical to a Non‑shared Subscription’s Topic Filter for the current Session, then it MUST replace that existing Subscription with a new Subscription [MQTT-3.8.4-3]. The Topic Filter in the new Subscription will be identical to that in the previous Subscription, although its Subscription Options could be different.